### ISSUE TITLE
Allow packaging arbitrary files

### ISSUE THREAD

**Issue by ericoporto** [adventuregamestudio/ags#1227]
My usecase is text files, like in-game books (`my-book.txt`). Maybe we could limit this if there's some dangerous use.

Basically, I would like to leverage AGS File.Open API but instead of reading from disk, I would like to read from non-exposed game resources (the own `.ags` game or maybe something like `.vox` packages) - I just don't would want people to **easily** read (`the-end.txt`) or tamper with my game, and currently the only way to do this is to write some script that generates AGS Script strings from text files in a script that is part of the game.

This would ease the process of revision and fixing of these text files that are part of the game for me.

**Comment by ivan-mogilko** [comment]
This was even done for a test before, but user who requested did not respond and feature got dropped. I'm not even fully certain if there's no ticket for this already somewhere...

Absolutely trivial to do in the engine, all you have to do is to use common AssetManager::OpenAsset function. That will automatically search in every registered resource location. There are certain rules for where non-sound files are searched (voxes are skipped), mostly done for compatibility reasons, as AGS still does not have a fully proper "virtual filesystem", but these may be worked around if necessary.

There are two questions remaining:

1. How to address these files. E.g. there's currently $INSTALLDIR$ which means "data directory", and you are only allowed to perform read operations there. We may simply use this, then it will mean simply "search in game data". The tag name would seem a bit off then though. *Or* we may introduce new tag, like $RESOURCE$ for searching strictly in packages, except that would contradict virtual filesystem concept.
2. How to package these. Unfortunately game project still has inconsistent rules about how resources are included. So guess if we work with that we need to either a) have a  special named folder for these files, where game compiler would take them automatically (I used this method in the mentioned test), b) create a new node in project tree and basically make it work similar to audioclips.

**Comment by ericoporto** [comment]
1. In my mind I had `$RESOURCE$` tag way. I know the engine magically picks up rooms if they are outside in some way but I don't know if this is related or not to things like `$INSTALLDIR$`. The need is read-only, yes!
2. A node in AGS project explorer like audio clips is the way I had on my mind. This also gives visual confirmation that things are going to get picked up.

My use case is really a bunch of in-game books and readable items that are easier to write and revise the text as a bunch of `.txt` files.

**Comment by ivan-mogilko** [comment]
I think it's worth to clarify, there are two ways to approach all this. One is to treat these user added files as a new type of resource with its own rules (how to package, how to open, etc).
Another is to treat them equaly to every other resource, and expand overall resource management as necessary. This is frankly what I'd prefer to do.

Then this splits into following tasks that may be reviewed and implemented separately:

1. Provide API for opening *any* game resource for reading, using existing engine's resource manager. In other words, it opens files exactly as engine would do if it wanted to read game data, following same rules and search logic.
If engine's virtual filesystem would be altered in the future, this API will automatically catch up to these changes.

2. Provide means for packaging selected files by the game compiler. If we had stand alone tools and open project format already that would not require anything extra at all, as ags package compiler would not care about file types and would simply pack what it's told by some script or command-line arguments.
The problem now is that game compilation is controlled inside the editor and it uses Game.agf to get list of files for most parts. Speech folder being the only exception (can't say that's a bad thing).
If we aim to stay consistent within *current* project rules, we may choose either - "any files found in special folder" or "files mentioned in Game.agf". Note that neither restricts from having these files mentioned in editor tree, but this will affect how UI works with these items.

3. Provide UI for configuring custom user files to be packaged along with the game. This is optional (speech already works without project tree), but its implementation depends on the p2.
If we package all files from special folder then UI should reflect actual filesystem. If we chose to have a list of files in Game.agf, then this works similar to sprites or audio files.

---

Something to also keep in mind here. In the past @morganwillcock proposed to change project structure to completely open format, where game project is basically what lies within the folder.
Don't want to delve into much details, but to answer most naturally raising argument - files and subfolders may be ignored same way as they are ignored by e.g. git: by having a "ignore" file with list of patterns inside. Alternatively - it may have an opposite rule where you should have a file with list of patterns telling which files and subfolders to *include* with the game.
Anyway, my point is, whatever we do, in my strong opinion, when adding new things we should consider what kind of system we want to have in the end. For instance, I'd like to have a proper virtual filesystem for engine, where it would support subfolders both on disk and in packages (regardless of how these are implemented in particular pack format), so that e.g. opening "data/file" or "music/clip" would work similarily if the file is on disk or in the packed file.

**Comment by rofl0r** [comment]
i agree that this functionality could be very useful (although i didnt understand if suggestion is to retrieve file from pack file or FS). a possible usecase is that a game may be compiled only once in editor, and from then on further functionality could be added via e.g. scripts compiled via agsutils or standalone agscc.

[in documentation for FileOpen() and friends](https://ags-archives.github.io/ags262/ags41.htm#topic62) we can read: 

> You should only use this with files which you previously wrote out with FileWrite. Do NOT use this function with any other files, even text files.

and 

> IMPORTANT: If you open the file for writing, then you can ONLY work with files in the game directory. You CANNOT use a path, so any filename with "\" or "/" in it will automatically be rejected, for security reasons.

issue 1 could be mitigated somewhat by using **FileReadRawChar (int handle)**:

>Reads a raw character from the input file HANDLE and returns it. This function allows you to read from files that weren't created by your game, however it is recommended for expert users only.

however reading only 1 char at a time in loop is really inefficient, and will take huge amount of time reading larger file (syscall overhead).

i propose to add functions `FileReadRaw(handle, char[], len)` and `FileWriteRaw(handle, char[], len)` which *try* to read/write len bytes if possible and return -1 for error or number of bytes read/written (0 if streamend is reached) just like the corresponding i/o functions in C/POSIX (write/read/fwrite/fread, etc).

additionally the restriction of not being able to write/read from subdirs should be removed, but for security/sandboxing purposes filenames shall not start with '/' or '\' and may not contain either of '/.' './' to prevent accessing files outside the game dir.

**Comment by ivan-mogilko** [comment]
> in documentation for FileOpen() and friends we can read:

> > You should only use this with files which you previously wrote out with FileWrite. Do NOT use this function with any other files, even text files.

Errrmm, thanks for bringing my attention to this, that is clearly redundant instruction, and some people already ignore that, for example, with IniFile module. This warning should be removed from the manual.

>>  If you open the file for writing, then you can ONLY work with files in the game directory. You CANNOT use a path, so any filename with "" or "/" in it will automatically be rejected, for security reasons.

This is also partially wrong now, as subdirectories are allowed for a while, and in write mode engine even creates all missing subdirectories. What is forbidden still is absolute paths for writing (I think reading might work, even though I'd wonder if that's a good idea or not).

> i propose to add functions FileReadRaw(handle, char[], len) and FileWriteRaw(handle, char[], len) 

That could be done as a separate feature, yes.

**Comment by ericoporto** [comment]
Found a topic about this feature: https://www.adventuregamestudio.co.uk/forums/index.php?topic=51595.0

**Comment by ivan-mogilko** [comment]
> Found a topic about this feature: https://www.adventuregamestudio.co.uk/forums/index.php?topic=51595.0

Ohh, that was so long ago. I might read this to refresh memories, but on another hand the newer idea seems more correct.

Anyway, *opening* the resource in the script might be easy, afaik the only additional requirement is to implement length restrictions to the filestream (as mentioned above), to prevent reading beyond the asset length when it's read from the bigger package. This may be done as a separate stream subclass, or temporarily as an option in regular filestream.
(I've been pondering on modifications to stream classes suggested by sonneveld, thinking maybe implement some ideas in 3.6.0, but that does not have to be done in a strict order)

Packing custom files is more complicated, as we need to integrate this within existing project setup, and add controls in the editor if we also want to see and manage these there. Which may be done as a separate task.

**Comment by ivan-mogilko** [comment]
Opening resource files in script: #1294

**Comment by ericoporto** [comment]
I think having something like a magic named folder like  `Data` or `ExtraAssets` in the project root could work for packaging and someday in the future adding a GUI in the Editor that reflects those in a project node so reading the assets names is easy and the feature is more discoverable (Speech gets a node too).
